File: Android/iEatAndroid/src/main/java/com/thoughtworks/ieat/activity/LoginActivity.scala
package com.thoughtworks.ieat.activity

import com.thoughtworks.ieat.Common._
import android.app.{ProgressDialog, Activity}
import android.os.{AsyncTask, Bundle}
import android.widget.{Toast, Button, EditText}
import android.content.{Context, Intent}
import android.view.View
import com.thoughtworks.ieat.{IEatApplication, activity, R}
import android.text.{Editable, TextWatcher}
import org.apache.http.client.methods.HttpPost
import org.apache.http.params.{BasicHttpParams, DefaultedHttpParams}
import org.apache.http.impl.client.DefaultHttpClient
import org.apache.http.{HttpStatus, HttpHost}
import com.thoughtworks.ieat.activity.utils.PropertyUtils
import android.view.View.OnFocusChangeListener
import scala.Array


class LoginActivity extends Activity {
  lazy val username = findViewById(R.id.username).asInstanceOf[EditText]
  lazy val password = findViewById(R.id.password).asInstanceOf[EditText]
  lazy val logInButton = findViewById(R.id.logInButton).asInstanceOf[Button]

  override def onCreate(savedInstanceState: Bundle) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.login)

//    logInButton.setOnClickListener {
//      (v: View) => {
//        val intent = new Intent(LoginActivity.this, classOf[DashboardActivity])
//        startActivity(intent)
//      }
//    }

    username.setOnFocusChangeListener(new OnFocusChangeListener {
      def onFocusChange(v: View, hasFocus: Boolean) {
        if (!hasFocus) {
          isCheckValidUsername(v.asInstanceOf[EditText].getText.toString)
        }
      }
    })

  }

  def isCheckValidUsername(username: String) : Boolean = {
    if (!username.endsWith("@thoughtworks.com")) {
      Toast.makeText(LoginActivity.this, "please input thoughtworks mail!", Toast.LENGTH_SHORT).show()
      return false
    }
    return true
  }

  def login(view : View) {

    val param: Array[String] = new Array[String](2)
    param(0) = username.getText.toString
    param(1) =  password.getText.toString

    if (!isCheckValidUsername(param.apply(0))) {
      return
    }

//    val loginTask: LoginTask = new LoginTask(this, param.apply(0), param.apply(1))
//    loginTask.execute()

    val loadingProgress = new ProgressDialog(this)
    onPreExecute(loadingProgress);

    val responseStatus = doInBackground(username.getText.toString, password.getText.toString)

    loadingProgress.dismiss()
    if (responseStatus == HttpStatus.SC_OK) {
      this.getApplication.asInstanceOf[IEatApplication].login(param.apply(0));
      Toast.makeText(this, "send succuessful", Toast.LENGTH_SHORT).show()
      val intent = new Intent(this, classOf[DashboardActivity])
      this.startActivity(intent)
    } else {
      Toast.makeText(this, responseStatus.toString, Toast.LENGTH_SHORT).show()
    }

  }

  def onPreExecute(loadingProgress: ProgressDialog) {
    loadingProgress.setMessage("login...")
    loadingProgress.setCanceledOnTouchOutside(false)
    loadingProgress.show();
  }

  def doInBackground(username : String, password : String) : Int = {
    val httpHost = new HttpHost(PropertyUtils.getServerHost(), PropertyUtils.getServerPort())

    val httpPost = new HttpPost("/users/sign_in")
    val httpParams = new BasicHttpParams()
    httpParams.setParameter("email", username)
    httpParams.setParameter("password", password)
    val httpClient = new DefaultHttpClient()
    httpClient.setParams(httpParams)
    try {
      val response = httpClient.execute(httpHost, httpPost)
      return response.getStatusLine.getStatusCode.toInt
    } catch {
      case exception: Exception => {
        Toast.makeText(this, exception.getMessage, Toast.LENGTH_SHORT).show()
        return 0.toInt
      }

    }

  }


}

class LoginTask(context: Context, username : String, password : String) extends AsyncTask[Void, Void, Int] {
  val loadingProgress = new ProgressDialog(context)

  override def onPreExecute() {
    loadingProgress.setMessage("login...")
    loadingProgress.setCanceledOnTouchOutside(false)
    loadingProgress.show();
  }

  override def doInBackground(p : Void*) : Int = {
    val httpHost = new HttpHost(PropertyUtils.getServerHost(), PropertyUtils.getServerPort())

    val httpPost = new HttpPost("/users/sign_in")
    val httpParams = new BasicHttpParams()
    httpParams.setParameter("email", username)
    httpParams.setParameter("password", password)
    val httpClient = new DefaultHttpClient()
    httpClient.setParams(httpParams)
    try {
      val response = httpClient.execute(httpHost, httpPost)
      return response.getStatusLine.getStatusCode.toInt
    } catch {
      case exception: Exception => {
        Toast.makeText(context, exception.getMessage, Toast.LENGTH_SHORT).show()
        return 0.toInt
      }

    }

  }

  override def onPostExecute(responseStatus : Int) {
    if (responseStatus == HttpStatus.SC_OK) {
      Toast.makeText(context, "send succuessful", Toast.LENGTH_SHORT).show()
      val intent = new Intent(context, classOf[DashboardActivity])
      context.startActivity(intent)
    } else {
      Toast.makeText(context, responseStatus.toString, Toast.LENGTH_SHORT).show()
    }
  }

}

File: Android/iEatAndroid/src/main/java/com/thoughtworks/ieat/activity/LoginActivity.java
package com.thoughtworks.ieat.activity;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;
import com.thoughtworks.ieat.IEatApplication;
import com.thoughtworks.ieat.R;
import com.thoughtworks.ieat.domain.AppHttpResponse;
import com.thoughtworks.ieat.domain.UserToken;
import com.thoughtworks.ieat.service.Server;

public class LoginActivity extends Activity {
    private EditText usernameView;
    private EditText passwordView;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.login);

        if (IEatApplication.currentUser() != null && IEatApplication.getToken() != null) {
            goToApp();
            finish();
            return;
        }

        usernameView = (EditText) findViewById(R.id.username);
        usernameView.setTypeface(Typeface.DEFAULT);
        passwordView = (EditText) findViewById(R.id.password);

        passwordView.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                login(null);
                return false;
            }
        });
    }

    private void goToApp() {
        Intent intent = new Intent(this, GroupListActivity.class);
        startActivity(intent);
        finish();
    }

    public void login(View view) {

        String username = usernameView.getText().toString();
        String pwd = passwordView.getText().toString();

        if (hasInputEmpty()) return;

        LoginAsyncTask loginTask = new LoginAsyncTask(this);
        loginTask.execute(username, pwd);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            IEatApplication.exit();
        }
        return super.onKeyDown(keyCode, event);
    }

    private boolean hasInputEmpty() {
        if (toastEmptyMessage(usernameView, R.string.login_username_label)) return true;
        if (toastEmptyMessage(passwordView, R.string.login_password_label)) return true;
        return false;
    }

    private boolean toastEmptyMessage(EditText editView, int viewLabelId) {
        if (editView.getText().toString().equals("")) {
            Toast.makeText(this, getString(viewLabelId) + getString(R.string.input_empty_message), Toast.LENGTH_SHORT).show();
            return true;
        }
        return false;
    }

    public void goToRegister(View view) {
        Intent intent = new Intent(this, RegisterActivity.class);
        startActivity(intent);
    }


    public class LoginAsyncTask extends AsyncTask<String, Void, AppHttpResponse<UserToken>> {

        private final Context context;
        private ProgressDialog loadingProgress;

        public LoginAsyncTask(Context context) {
            this.context = context;
        }

        public void onPreExecute() {
            loadingProgress = new ProgressDialog(context);
            loadingProgress.setMessage("login...");
            loadingProgress.setCanceledOnTouchOutside(false);
            loadingProgress.show();
        }

        public AppHttpResponse<UserToken> doInBackground(String... params) {
            AppHttpResponse appHttpResponse = Server.signIn(params[0], params[1]);
            return appHttpResponse;
        }

        public void onPostExecute(AppHttpResponse<UserToken> response) {
            loadingProgress.dismiss();
            if (response.isSuccessful()) {
                Intent intent = new Intent(context, GroupListActivity.class);
                startActivity(intent);
                finish();
            } else {
                Toast.makeText(context, response.getErrorMessage(), Toast.LENGTH_SHORT).show();
            }
        }
    }
}
